services/group_service.go

```diff
@@ -0,0 +1,153 @@
+
+	return group, nil
+}
+
+func (gs *GroupService) UpdateGroup(ctx context.Context, group *datastore.Group, update *models.Group) (*datastore.Group, error) {
+	groupName := update.Name
+	err := util.Validate(update)
+	if err != nil {
+		log.WithError(err).Error("failed to validate group update")
+		return nil, NewServiceError(http.StatusBadRequest, err)
+	}
+
+	group.Name = groupName
+	group.Config = &update.Config
+	if !util.IsStringEmpty(update.LogoURL) {
+		group.LogoURL = update.LogoURL
+	}
+
+	err = gs.groupRepo.UpdateGroup(ctx, group)
+	if err != nil {
+		log.WithError(err).Error("failed to to update group")
+		return nil, NewServiceError(http.StatusInternalServerError, errors.New("an error occurred while updating Group"))
+	}
+
+	return group, nil
+}
+
+func (gs *GroupService) GetGroups(ctx context.Context, filter *datastore.GroupFilter) ([]*datastore.Group, error) {
+	groups, err := gs.groupRepo.LoadGroups(ctx, filter)
+	if err != nil {
+		log.WithError(err).Error("failed to load groups")
+		return nil, NewServiceError(http.StatusInternalServerError, errors.New("an error occurred while fetching Groups"))
+	}
+
+	for _, group := range groups {
+		err = gs.FillGroupStatistics(ctx, group)
+		if err != nil {
+			log.WithError(err).Errorf("failed to fill statistics of group %s", group.UID)
+		}
+	}
+	return groups, nil
+}
+
+func (gs *GroupService) FillGroupStatistics(ctx context.Context, g *datastore.Group) error {
+	appCount, err := gs.appRepo.CountGroupApplications(ctx, g.UID)
+	if err != nil {
+		log.WithError(err).Error("failed to count group applications")
+		return NewServiceError(http.StatusInternalServerError, errors.New("failed to count group statistics"))
+	}
+
+	msgCount, err := gs.eventRepo.CountGroupMessages(ctx, g.UID)
+	if err != nil {
+		log.WithError(err).Error("failed to count group messages")
+		return NewServiceError(http.StatusInternalServerError, errors.New("failed to count group statistics"))
+	}
+
+	g.Statistics = &datastore.GroupStatistics{
+		MessagesSent: msgCount,
+		TotalApps:    appCount,
+	}
+	return nil
+}
+
+func (gs *GroupService) DeleteGroup(ctx context.Context, id string) error {
+	err := gs.groupRepo.DeleteGroup(ctx, id)
+	if err != nil {
+		log.WithError(err).Error("failed to delete group")
+		return NewServiceError(http.StatusInternalServerError, errors.New("failed to delete group"))
+	}
+
+	// TODO(daniel,subomi): is returning http error necessary for these? since the group itself has been deleted
+	err = gs.appRepo.DeleteGroupApps(ctx, id)
+	if err != nil {
+		log.WithError(err).Error("failed to delete group apps")
+		return NewServiceError(http.StatusInternalServerError, errors.New("failed to delete group apps"))
+	}
+
+	err = gs.eventRepo.DeleteGroupEvents(ctx, id)
+	if err != nil {
+		log.WithError(err).Error("failed to delete group events")
+		return NewServiceError(http.StatusInternalServerError, errors.New("failed to delete group events"))
```

Comment:
Can we use this opportunity to change all the occurrences of this status code (for the files in this PR alone)  
```suggestion
		return NewServiceError(http.StatusBadRequest, errors.New("failed to delete group events"))
```

Reply:
Point taken. I will implement this.